qa: pre-procesar reportes JUnit/Maestro a resumen estructurado

## Contexto

Hoy el agente qa, después de correr los tests, **lee los XMLs de JUnit y los logs de Maestro a mano con Read**, recorriendo cientos de líneas para identificar:

- Cuántos tests pasaron / fallaron / se saltearon.
- Qué tests fallaron y por qué (stack trace).
- Tiempo total y por test.
- Si hay flakiness (test que pasa solo a veces).

Eso lo hace bien, pero quema cache_read leyendo XMLs grandes y output de Gradle (\~80-200 líneas por test) que en su mayoría son ruido (timestamps, configuración Gradle, stdout de tests que pasaron).

## Propuesta

Crear `qa/scripts/qa-summarize-results.js` que:

1. Recibe paths a `qa/build/test-results/test/` y/o `qa/recordings/maestro-results.xml` y/o `app/composeApp/build/reports/tests/desktopTest/`.
2. Parsea los XMLs JUnit (xml2js o regex simple) y los reports Maestro.
3. Devuelve JSON estructurado:

```json
{
  \"summary\": {
    \"total\": 47,
    \"passed\": 45,
    \"failed\": 2,
    \"skipped\": 0,
    \"duration_ms\": 132450,
    \"platforms\": [\"api\", \"android\"]
  },
  \"failures\": [
    {
      \"platform\": \"api\",
      \"class\": \"ApiSignInE2ETest\",
      \"name\": \"signin con datos válidos responde 200\",
      \"duration_ms\": 4502,
      \"reason\": \"AssertionError: status was 500, expected 2xx\",
      \"stack_top\": \"at ar.com.intrale.e2e.ApiSignInE2ETest...\"
    }
  ],
  \"slow_tests\": [...],
  \"warnings\": [...]
}
```

4. El LLM solo lee este JSON y redacta el reporte/veredicto sobre datos limpios — no relee XMLs.

## Ahorro estimado

- Lectura de reportes: 5-15K tokens cache_read → \~500 tokens (\-95% solo en esa fase).
- Total estimado del agente: \-15-20% adicional sobre lo que se ahorra con los otros 3 issues.
- Bonus: el JSON es archivable como `qa/evidence/<issue>/qa-results-summary.json` y queda como artefacto trazable.

## Aceptación

- [ ] `qa/scripts/qa-summarize-results.js` creado con parser JUnit + Maestro.
- [ ] Self-test: corre contra `qa/build/test-results/` real (con tests pasados o fallidos) y emite JSON válido.
- [ ] `SKILL.md` actualizado: paso 3 (\"Analizar resultados\") invoca el script en lugar de leer XMLs a mano.
- [ ] El JSON se persiste en `qa/evidence/<issue>/qa-results-summary.json` como artefacto.
- [ ] Defensive programming: si no hay reportes, devuelve `{summary:{total:0...}, warnings:[\"sin_reportes\"]}` sin tirar excepción.

## Skill responsable

`pipeline-dev` (extracción de script, sin impacto en producto → `qa:skipped`).

--- a/qa/scripts/qa-summarize-results.js
+++ b/qa/scripts/qa-summarize-results.js
@@ -1,0 +1,253 @@
+#!/usr/bin/env node
+// qa-summarize-results.js — Pre-procesa reportes JUnit/Maestro a resumen estructurado.
+//
+// Uso:
+//   node qa/scripts/qa-summarize-results.js [--out <path>]
+//
+// Sin argumentos: localiza reportes en paths estándar y emite JSON por stdout.
+// Con --out: además persiste el JSON en el path indicado.
+//
+// Reportes que considera:
+//   qa/build/test-results/test/*.xml                       (API JUnit)
+//   app/composeApp/build/test-results/desktopTest/*.xml    (Desktop JUnit)
+//   qa/recordings/maestro-results.xml                       (Maestro)
+//
+// Salida JSON:
+//   {
+//     "summary": { total, passed, failed, skipped, duration_ms, platforms: [...] },
+//     "failures": [ { platform, class, name, duration_ms, reason, stack_top } ],
+//     "slow_tests": [ { platform, class, name, duration_ms } ],   // top 5 si > 5s
+//     "warnings": [ ... ],
+//     "sources": { junit_api: [...], junit_desktop: [...], maestro: <path|null> }
+//   }
+// Exit code: 0 siempre (ausencia de reportes se reporta en warnings).
+
+'use strict';
+
+const fs = require('fs');
+const path = require('path');
+
+const SLOW_THRESHOLD_MS = 5000;
+const SLOW_TOP_N = 5;
+
+function readSafe(p) {
+  try {
+    return fs.readFileSync(p, 'utf8');
+  } catch {
+    return null;
+  }
+}
+
+function listXml(dir) {
+  if (!fs.existsSync(dir)) return [];
+  try {
+    return fs
+      .readdirSync(dir)
+      .filter((f) => f.endsWith('.xml'))
+      .map((f) => path.join(dir, f));
+  } catch {
+    return [];
+  }
+}
+
+function decodeXmlEntities(s) {
+  if (!s) return s;
+  return s
+    .replace(/&lt;/g, '<')
+    .replace(/&gt;/g, '>')
+    .replace(/&quot;/g, '"')
+    .replace(/&apos;/g, "'")
+    .replace(/&amp;/g, '&');
+}
+
+// Parsea un XML JUnit. Acepta <testsuite> raíz o anidados en <testsuites>.
+function parseJUnit(xml, platform, sourcePath) {
+  const result = { tests: [], parseErrors: [] };
+  if (!xml) return result;
+
+  const testcaseRegex = /<testcase\b([^>]*?)(\/>|>([\s\S]*?)<\/testcase>)/g;
+  const attrRegex = /(\w+)\s*=\s*"([^"]*)"/g;
+
+  let match;
+  while ((match = testcaseRegex.exec(xml)) !== null) {
+    const attrs = {};
+    let am;
+    const attrStr = match[1];
+    while ((am = attrRegex.exec(attrStr)) !== null) {
+      attrs[am[1]] = am[2];
+    }
+    const inner = match[3] || '';
+
+    const className = attrs.classname || attrs.class || '';
+    const name = attrs.name || '';
+    const durationMs = attrs.time ? Math.round(parseFloat(attrs.time) * 1000) : 0;
+
+    let status = 'passed';
+    let reason = null;
+    let stackTop = null;
+
+    const failureMatch = inner.match(/<(failure|error)\b([^>]*?)(\/>|>([\s\S]*?)<\/\1>)/);
+    if (failureMatch) {
+      status = 'failed';
+      const failAttrs = {};
+      let fa;
+      const fAttrRegex = /(\w+)\s*=\s*"([^"]*)"/g;
+      while ((fa = fAttrRegex.exec(failureMatch[2])) !== null) {
+        failAttrs[fa[1]] = fa[2];
+      }
+      reason = decodeXmlEntities(failAttrs.message || failAttrs.type || 'unknown');
+      const body = decodeXmlEntities(failureMatch[4] || '');
+      const firstLine = body.split('\n').find((l) => l.trim().length > 0) || '';
+      stackTop = firstLine.trim().slice(0, 240);
+    } else if (/<skipped\b/.test(inner)) {
+      status = 'skipped';
+    }
+
+    result.tests.push({
+      platform,
+      class: className,
+      name,
+      status,
+      duration_ms: durationMs,
+      reason,
+      stack_top: stackTop,
+      source: sourcePath,
+    });
+  }
+
+  return result;
+}
+
+function buildSummary(tests, platforms) {
+  const summary = {
+    total: tests.length,
+    passed: 0,
+    failed: 0,
+    skipped: 0,
+    duration_ms: 0,
+    platforms: Array.from(new Set(platforms)).sort(),
+  };
+  for (const t of tests) {
+    summary.duration_ms += t.duration_ms || 0;
+    if (t.status === 'passed') summary.passed++;
+    else if (t.status === 'failed') summary.failed++;
+    else if (t.status === 'skipped') summary.skipped++;
+  }
+  return summary;
+}
+
+function buildSlowTests(tests) {
+  return tests
+    .filter((t) => (t.duration_ms || 0) >= SLOW_THRESHOLD_MS)
+    .sort((a, b) => b.duration_ms - a.duration_ms)
+    .slice(0, SLOW_TOP_N)
+    .map(({ platform, class: cls, name, duration_ms }) => ({
+      platform,
+      class: cls,
+      name,
+      duration_ms,
+    }));
+}
+
+function parseArgs(argv) {
+  const args = { out: null };
+  for (let i = 2; i < argv.length; i++) {
+    if (argv[i] === '--out' && argv[i + 1]) {
+      args.out = argv[++i];
+    }
+  }
+  return args;
+}
+
+function main() {
+  const args = parseArgs(process.argv);
+  const warnings = [];
+
+  const apiXml = listXml('qa/build/test-results/test');
+  const desktopXml = listXml('app/composeApp/build/test-results/desktopTest');
+  const maestroXml = fs.existsSync('qa/recordings/maestro-results.xml')
+    ? 'qa/recordings/maestro-results.xml'
+    : null;
+
+  if (apiXml.length === 0 && desktopXml.length === 0 && !maestroXml) {
+    warnings.push('no_reports_found');
+  }
+
+  const allTests = [];
+  const platforms = [];
+  const failures = [];
+
+  for (const p of apiXml) {
+    const xml = readSafe(p);
+    if (!xml) {
+      warnings.push(`unreadable: ${p}`);
+      continue;
+    }
+    const r = parseJUnit(xml, 'api', p);
+    allTests.push(...r.tests);
+    if (r.tests.length > 0) platforms.push('api');
+  }
+
+  for (const p of desktopXml) {
+    const xml = readSafe(p);
+    if (!xml) {
+      warnings.push(`unreadable: ${p}`);
+      continue;
+    }
+    const r = parseJUnit(xml, 'desktop', p);
+    allTests.push(...r.tests);
+    if (r.tests.length > 0) platforms.push('desktop');
+  }
+
+  if (maestroXml) {
+    const xml = readSafe(maestroXml);
+    if (!xml) {
+      warnings.push(`unreadable: ${maestroXml}`);
+    } else {
+      const r = parseJUnit(xml, 'android', maestroXml);
+      allTests.push(...r.tests);
+      if (r.tests.length > 0) platforms.push('android');
+    }
+  }
+
+  for (const t of allTests) {
+    if (t.status === 'failed') {
+      failures.push({
+        platform: t.platform,
+        class: t.class,
+        name: t.name,
+        duration_ms: t.duration_ms,
+        reason: t.reason,
+        stack_top: t.stack_top,
+      });
+    }
+  }
+
+  const result = {
+    summary: buildSummary(allTests, platforms),
+    failures,
+    slow_tests: buildSlowTests(allTests),
+    warnings,
+    sources: {
+      junit_api: apiXml,
+      junit_desktop: desktopXml,
+      maestro: maestroXml,
+    },
+  };
+
+  const json = JSON.stringify(result, null, 2);
+
+  if (args.out) {
+    try {
+      fs.mkdirSync(path.dirname(args.out), { recursive: true });
+      fs.writeFileSync(args.out, json);
+    } catch (err) {
+      warnings.push(`failed to write ${args.out}: ${err.message}`);
+    }
+  }
+
+  process.stdout.write(json + '\n');
+  process.exit(0);
+}
+
+main();

